### PR TITLE
feat: [kubelet/cAdvisor metrics package] add NodeMonitoring ScrapeConfig logic

### PR DESCRIPTION
**Pull request by macxamin**
All changes are in types.go and types_test.go. The rest of the files are generated via `make regen`

types.go changes:
1. Add functions needed to support NodeMonitoring. Will refactor code in the future to reuse functions from PodMonitoring.

types_test.go changes:
1. Add selector to `TestPodMonitoring_ScrapeConfig` and `TestClusterPodMonitoring_ScrapeConfig` since `relabelingsForSelector` was modified. 
2. Added a tests to validate and create NodeMonitoring Scrape configs.

### CODE REVIEW

File: pkg/operator/apis/monitoring/v1/types.go [hunk 1]
```diff
@@ -1405,6 +1417,201 @@ type NodeMonitoring struct {
 	Spec NodeMonitoringSpec `json:"spec"`
 }
 
+// NodeScrapeEndpoint specifies a node's Prometheus metrics endpoint to scrape.
+type NodeScrapeEndpoint struct {
```
Comment (bwplotka) [inline pkg/operator/apis/monitoring/v1/types.go:1421]: * Hm.. I am missing "port" 🤔 
* Somehow interval is required?
image
* Scheme (you can totally use TLS) and basic HTTPClientConfig without secrets would be nice for a start 🤔 (to at least skip verify?)
* Params we could bring back as well, there is little harm

File: pkg/operator/apis/monitoring/v1/types.go [hunk 2]
```diff
@@ -1405,6 +1417,201 @@ type NodeMonitoring struct {
...
+}
+
+func (nm *NodeMonitoring) endpointScrapeConfig(index int, projectID, location, cluster string) (*promconfig.ScrapeConfig, error) {
+	// Filter targets that belong to selected nodes.
+	relabelCfgs, err := relabelingsForSelector(nm.Spec.Selector, nm)
+	if err != nil {
+		return nil, err
+	}
+
+	relabelCfgs = append(relabelCfgs,
```
Comment (bwplotka) [inline pkg/operator/apis/monitoring/v1/types.go:1484]: idea: hmmm I wonder if there's no opportunity to reuse some code between Pod/Cluster monitorings. Not a blocker (:

**Review by bwplotka — COMMENTED** [review]
Thanks! Some comments around left out fields in NodeMonitoring, otherwise fine 👍🏽

File: pkg/operator/apis/monitoring/v1/types.go (hunk 1 above)
Comment (macxamin) [reply pkg/operator/apis/monitoring/v1/types.go:1421]: - I removed port, since it's not part of the [prometheus metadata labels for nodes. ](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#node) Do you see users wanting to user port?
- If IIUC the Kubelet serves metrics over HTTPs, so I removed it to reduce the chance of a user error and simplify the code. I guess we can return error if it's not https. The kubelet has it's ca. So we need the HTTPClientConfig to look like this:
image
Should we expose HTTPClientConfig and just block those fields?
- Yeah, I can bring it back.

Comment (bwplotka) [reply pkg/operator/apis/monitoring/v1/types.go:1421]: Got it, sounds like you want to tailor it to kubelet. However given "NodeMonitoring" semantics that might be surprising 🤔 I think I would be ok to start with tailored kubelet approach you propose, wonder what's @pintohutch opinion.

Comment (pintohutch) [reply pkg/operator/apis/monitoring/v1/types.go:1421]: Discussed with @macxamin offline - IIUC we want to keep port number and http as options. Port name however isn't applicable for the node as it's not in the [spec](https://pkg.go.dev/k8s.io/api/core/v1#NodeDaemonEndpoints).

**Comment by bwplotka** [comment]
Let us know once it's rdy for a second review iteration @macxamin ! 💪🏽

File: pkg/operator/apis/monitoring/v1/types.go (hunk 1 above)
Comment (macxamin) [reply pkg/operator/apis/monitoring/v1/types.go]: Yes, I will add port logic in a future PR. I want to get nodeMonitoring to a working state and then I can make refinements to it.

**Comment by macxamin** [comment]
> Let us know once it's rdy for a second review iteration @macxamin ! 💪🏽

@bwplotka Ready!

File: pkg/operator/apis/monitoring/v1/types.go [hunk 3]
```diff
@@ -1406,6 +1418,189 @@ type NodeMonitoring struct {
...
+		},
+		TLSConfig: config.TLSConfig{
+			CAFile: "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt",
+		},
+	}
+	return extractScrapeParams(nm.GetKey(), ep, httpCfg, relabelCfgs, discoveryCfgs, nm.Spec.Limits)
+}
+
+// extractScrapeParams extracts the scrape parameters of a NodeMonitoring ScrapeEndpoint.
+func extractScrapeParams(id string, ep ScrapeEndpoint, httpCfg config.HTTPClientConfig, relabelCfgs []*relabel.Config, discoveryCfgs discovery.Configs, limits *ScrapeLimits) (*promconfig.ScrapeConfig, error) {
```
Comment (bwplotka) [inline pkg/operator/apis/monitoring/v1/types.go:1540]: hm... something could be improved here. Is it specific to NodeMonitoring (as comment suggests?). I don't think it is at the moment.

Also the function name is pretty generic here.. maybe simply `newPromScrapeConfig` is better? Or `endpointScrapeConfigForNM`, or `func (nm *NodeMonitoring) newPromScrapeConf`? Essentially it feels you tried to mimic https://github.com/GoogleCloudPlatform/prometheus-engine/blob/69d276c973b8e4f00531a3366e4140998a10da4d/pkg/operator/apis/monitoring/v1/types.go#L616 function, but this function is only only used once. 

In this state you could reuse `extractScrapeParam` function in https://github.com/GoogleCloudPlatform/prometheus-engine/blob/69d276c973b8e4f00531a3366e4140998a10da4d/pkg/operator/apis/monitoring/v1/types.go#L616, but not sure if want reuse ScrapeEndpoint for NodeMonitoring, what's the plan there, given my comments above?

**Review by bwplotka — COMMENTED** [review]
Thanks! Some questions/suggestions (:

File: pkg/operator/apis/monitoring/v1/types.go [hunk 4]
```diff
@@ -1406,6 +1418,189 @@ type NodeMonitoring struct {
...
+	var metricRelabelCfgs []*relabel.Config
+	for _, r := range ep.MetricRelabeling {
+		rcfg, err := convertRelabelingRule(r)
+		if err != nil {
+			return nil, err
+		}
+		metricRelabelCfgs = append(metricRelabelCfgs, rcfg)
+	}
+
+	// return interval, timeout, path, metricRelabelCfgs, nil
```
[the remark this answers is not in the document]
Comment (macxamin) [reply pkg/operator/apis/monitoring/v1/types.go:1569]: good catch

File: pkg/operator/apis/monitoring/v1/types.go [hunk 5]
```diff
@@ -1406,6 +1418,189 @@ type NodeMonitoring struct {
...
+				{
+					Role:  discoverykube.RoleNode,
+					Field: fmt.Sprintf("metadata.name=$(%s)", EnvVarNodeName),
+				},
+			},
+		},
+	}
+
+	ep := nm.Spec.Endpoints[index]
+	// TODO(macxamin) Add support for int port. String ports are not supported by the Kubernetes spec: https://pkg.go.dev/k8s.io/api/core/v1#NodeDaemonEndpoints.
```
[the remark this answers is not in the document]
Comment (macxamin) [reply pkg/operator/apis/monitoring/v1/types.go:1455]: Discussed offline, will add via feature branch

File: pkg/operator/apis/monitoring/v1/types.go [hunk 6]
```diff
@@ -1406,6 +1418,189 @@ type NodeMonitoring struct {
...
+	if ep.Port.StrVal != "" || ep.Port.IntVal != 0 {
+		return nil, errors.New("port not supported")
+	}
+	// TODO(macxamin) Add support for HTTPClientConfig. Add it for the user for now.
```
[the remark this answers is not in the document]
Comment (macxamin) [reply pkg/operator/apis/monitoring/v1/types.go:1459]: Discussed offline, will add via feature branch

File: pkg/operator/apis/monitoring/v1/types.go (hunk 3 above)
Comment (macxamin) [reply pkg/operator/apis/monitoring/v1/types.go]: done. PTAL.

File: pkg/operator/apis/monitoring/v1/types.go [hunk 7]
```diff
@@ -1374,13 +1337,52 @@ type RulesStatus struct {
 	// TODO: add status information.
 }
 
+// CommonScrapeEndpointSpec
```
Comment (bwplotka) [inline pkg/operator/apis/monitoring/v1/types.go:1340]: Some comment left from older impl?

File: pkg/operator/apis/monitoring/v1/types.go [hunk 8]
```diff
@@ -1374,13 +1337,52 @@ type RulesStatus struct {
...
+	timeout          string
+	params           map[string][]string
+	metricRelabeling []RelabelingRule
+	relabelCfgs      []*relabel.Config
+	httpCfg          config.HTTPClientConfig
+	discoveryCfgs    discovery.Configs
+	limits           *ScrapeLimits
+}
+
+// ScrapeNodeEndpoint specifies a Prometheus metrics endpoint on a node to scrape.
```
Comment (bwplotka) [inline pkg/operator/apis/monitoring/v1/types.go:1355]: Could you summarize here, the difference between ScrapeEndpoint and this (e.g TLS fields are not present)?  (for users & future devs)

File: pkg/operator/apis/monitoring/v1/types.go [hunk 9]
```diff
@@ -1374,13 +1337,52 @@ type RulesStatus struct {
 	// TODO: add status information.
 }
 
+// CommonScrapeEndpointSpec
+type endpointParams struct {
```
Comment (bwplotka) [inline pkg/operator/apis/monitoring/v1/types.go:1341]: Hm, sounds like this is only for one function use (that we use twice). Perhaps we could keep ScrapeEndpoint in that function as param and kill this struct (and make sure we transition from ScrapeNodeEndpoint to ScrapeEndpoint for that one function)? Sounds like a similar choice, just less code (I understand roughly ScrapeEndpoint >= ScrapeNodeEndpoint in scope)

**Review by bwplotka — APPROVED** [review]
Thank you! Two suggestions, otherwise LGTM!

File: pkg/operator/apis/monitoring/v1/types.go (hunk 9 above)
Comment (macxamin) [reply pkg/operator/apis/monitoring/v1/types.go:1341]: Great idea!
